service/collector/azure_api_metrics.go

```diff
@@ -0,0 +1,103 @@
+package collector
+
+import (
+	"sync"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+}
+
+type AzureAPIMetricsCollector struct {
+	logger micrologger.Logger
+
+	counters   map[string]*prometheus.CounterVec
+	histograms map[string]*prometheus.HistogramVec
+
+	mutex *sync.Mutex
+}
+
+func NewAzureAPIMetricsCollector(config Config) (*AzureAPIMetricsCollector, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	c := AzureAPIMetricsCollector{
+		logger: config.Logger,
+
+		counters:   map[string]*prometheus.CounterVec{},
+		histograms: map[string]*prometheus.HistogramVec{},
+		mutex:      &sync.Mutex{},
+	}
+
+	return &c, nil
+}
+
+func (c *AzureAPIMetricsCollector) Describe(ch chan<- *prometheus.Desc) error {
+	for _, c := range c.counters {
+		c.Describe(ch)
+	}
+
+	for _, h := range c.histograms {
+		h.Describe(ch)
+	}
+
+	return nil
+}
+
+func (c *AzureAPIMetricsCollector) Collect(ch chan<- prometheus.Metric) error {
+	for _, c := range c.counters {
+		c.Collect(ch)
+	}
+
+	for _, h := range c.histograms {
+		h.Collect(ch)
+	}
+
+	return nil
+}
+
+func (c *AzureAPIMetricsCollector) GetCounterVec(opts prometheus.Opts, labelNames []string) *prometheus.CounterVec {
+	k := opts.Namespace + "/" + opts.Name
+	counter, exists := c.counters[k]
+	if !exists {
+		c.mutex.Lock()
```

Comment:
why not use the mutex at the beginning of the function and only have one `if` statement?

Reply:
It's only needed once when creating the object.